internal/catalogmetadata/cache/cache.go

```diff
@@ -0,0 +1,152 @@
+	}
+}
+
+// cacheData holds information about a catalog
+// other than it's contents that is used for
+// making decisions on when to attempt to refresh
+// the cache.
+type cacheData struct {
+	LastModified string
+	Generation   int64
+	ResolvedRef  string
+}
+
+// FilesystemCache is a cache that
+// uses the local filesystem for caching
+// catalog contents. It will fetch catalog
+// contents if the catalog does not already
+// exist in the cache.
+type FilesystemCache struct {
+	mutex                  sync.RWMutex
+	cachePath              string
+	tripper                http.RoundTripper
+	cacheDataByCatalogName map[string]cacheData
+}
+
+// FetchCatalogContents will fetch the contents for the provided
+// Catalog from the filesystem. If the provided Catalog has not yet
+// been cached, it will make a GET request to the Catalogd HTTP server
+// to get the Catalog contents and cache them. The cache will be updated
+// automatically if a Catalog is noticed to have a different generation
+// and resolved image reference.
+// The Catalog provided to this function is expected to:
+// - Be non-nil
+// - Have a Catalog.Status.Conditions entry with type Unpacked and status True
+// - Have a non-nil Catalog.Status.ResolvedSource.Image
+// This ensures that we are only attempting to fetch catalog contents for Catalog
+// resources that have been successfully reconciled, unpacked, and are being served.
+// These requirements help ensure that we can rely on status conditions to determine
+// when to issue a request to update the cached Catalog contents.
+func (fsc *FilesystemCache) FetchCatalogContents(ctx context.Context, catalog *catalogd.Catalog) (io.ReadCloser, error) {
+	fsc.mutex.Lock()
+	defer fsc.mutex.Unlock()
+
+	shouldUpdateCache := false
+	data, cached := fsc.cacheDataByCatalogName[catalog.Name]
+	if cached {
+		// we should update the cache if the catalog generation has changed
+		// AND the status has been updated to reflect a new image ref
+		shouldUpdateCache = catalog.Generation != data.Generation && catalog.Status.ResolvedSource.Image.Ref != data.ResolvedRef
+	}
+
+	cacheDir := filepath.Join(fsc.cachePath, catalog.Name)
+	cacheFilePath := filepath.Join(cacheDir, "data.json")
+
+	if !cached || shouldUpdateCache {
+		err := os.MkdirAll(cacheDir, os.ModePerm)
+		if err != nil {
+			return nil, fmt.Errorf("error creating cache directory for Catalog %q: %s", catalog.Name, err)
+		}
+
+		// TODO: When a catalogd release containing https://github.com/operator-framework/catalogd/pull/168
+		// is made this should be updated in favor of parsing `Catalog.Status` for the
+		// appropriate URL for fetching catalog contents.
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf(catalogdOnClusterURLTemplate, catalogdOnClusterBaseURL, catalog.Name), nil)
+		if err != nil {
+			return nil, fmt.Errorf("error forming request: %s", err)
+		}
+
+		// If we are updating the cache, make sure there are actually changes
```

Comment:
If we can be certain of our intent by looking at the `*catalog.Catalog` status for the image ref, do we need these headers?

Reply:
Good point - removed in https://github.com/operator-framework/operator-controller/pull/411/commits/fe924c2c2f9c686d3720bb930d42f01590409df1